timeflux/nodes/window.py

```diff
@@ -16,33 +16,36 @@ def __init__(self, length, step=None):
                 If None (the default), the step will be set to the window duration.
                 If 0, the data will be sent as soon as it is available.
         """
-
-        if step is None: step = length
+        if step is None:
+            step = length
+        if step > length:
+            raise ValueError('Step of window should be smaller or equal to length. ')
         self._length = pd.Timedelta(seconds=length)
         self._step = pd.Timedelta(seconds=step)
         self._buffer = None
-        self._updated = pd.Timestamp(0)
 
     def update(self):
         # Return immediately if we don't have any data
         if not self.i.ready():
             return
+        # Sanity check
+        if not self.i.data.index.is_monotonic:
+            self.logger.warning('Window received data with not monotonic index.')
+
         # Append new data
         if self._buffer is None:
             self._buffer = self.i.data
         else:
             self._buffer = self._buffer.append(self.i.data)
-        # Time range
-        high = self._buffer.index[-1]
-        low = high - self._length
-        # Make sure we have enough data
-        if self._buffer.index[0] <= low:
-            # Step
-            if high - self._updated >= self._step:
-                # Clear old data
-                self._buffer = self._buffer[self._buffer.index > low]
-                # Remember the last time the buffer was updated
-                self._updated = high
-                # Output
-                self.o.data = self._buffer
-                self.o.meta = self.i.meta
+        # Calculate if we are ready to window the current buffer
+        low = self._buffer.index[0]
+        high = low + self._length
+        if self._buffer.index[-1] >= high:
+            self.o.data = self._buffer[self._buffer.index < high]
+            self.o.meta = self.o.meta
+            self._buffer = self._buffer[self._buffer.index >= low + self._step]
+
+        if not self._buffer.empty and (self._buffer.index[-1] - self._buffer.index[0]) > self._length:
+            self.logger.warning('This node is falling behind: it is receiving '
```

Comment:
If we are falling behind, we should warn AND truncate the buffer to avoid memory leaks.

Reply:
ok done